File: src/lib/rusto/src/base_should.rs
use crate::base_action::BaseAction;
use crate::types::IbddInAny;
use std::collections::HashMap;

/// BaseShould extends BaseAction for TDT pattern.
/// Processes each row in table-driven testing.
pub struct BaseShould<I: IbddInAny> {
    inner: BaseAction<I>,
    current_row: Vec<Box<dyn std::any::Any>>,
    row_index: i32,
}

impl<I: IbddInAny> BaseShould<I> {
    pub fn new(name: String, should_cb: Box<dyn Fn(I::Iselection) -> I::Then>) -> Self {
        Self {
            inner: BaseAction::new(name, should_cb),
            current_row: Vec::new(),
            row_index: -1,
        }
    }
    
    pub fn add_artifact(&mut self, path: String) {
        self.inner.add_artifact(path);
    }
    
    pub fn to_obj(&self) -> HashMap<String, Box<dyn std::any::Any>> {
        self.inner.to_obj()
    }
    
    // Set current row data
    pub fn set_row_data(&mut self, row_index: i32, row_data: Vec<Box<dyn std::any::Any>>) {
        self.row_index = row_index;
        self.current_row = row_data;
    }
    
    // Process the current row
    pub async fn process_row(
        &mut self,
        store: I::Istore,
        test_resource: &dyn std::any::Any,
    ) -> Result<I::Istore, String> {
        self.inner.test(store, test_resource).await
    }
    
    // Alias test to should for TDT pattern
    pub async fn should(
        &mut self,
        store: I::Istore,
        test_resource: &dyn std::any::Any,
    ) -> Result<I::Istore, String> {
        self.process_row(store, test_resource).await
    }
}

File: src/lib/tiposkripto/src/verbs/tdt/BaseShould.ts
import { TestTypeParams_any } from "../../CoreTypes.js";

/**
 * BaseShould for TDT pattern - independent implementation
 * Processes each row in table-driven testing.
 */
export class BaseShould<I extends TestTypeParams_any> {
  name: string;
  shouldCB: (xyz: I["iselection"]) => I["action"];
  currentRow: any[] = [];
  rowIndex: number = -1;
  error: Error | null = null;
  status: boolean | undefined;

  constructor(name: string, shouldCB: (xyz: I["iselection"]) => I["action"]) {
    this.name = name;
    this.shouldCB = shouldCB;
  }

  // Set current row data
  setRowData(rowIndex: number, rowData: any[]) {
    this.rowIndex = rowIndex;
    this.currentRow = rowData;
  }

  // Process the current row
  async processRow(
    actualResult: any,
    testResourceConfiguration: any,
    artifactory?: any,
  ) {
    try {
      let success = false;
      if (typeof this.shouldCB === 'function') {
        // The shouldCB expects a store, but we pass actualResult as store
        const result = await this.shouldCB(actualResult);
        success = !!result;
      } else {
        // shouldCB is the expected value
        success = actualResult === this.shouldCB;
      }
      this.status = success;
      return success;
    } catch (e: any) {
      this.status = false;
      this.error = e;
      throw e;
    }
  }

  toObj() {
    return {
      name: this.name,
      status: this.status,
      error: this.error ? `${this.error.name}: ${this.error.message}` : null,
      rowIndex: this.rowIndex,
      currentRow: this.currentRow,
      pattern: 'tdt'
    };
  }
}

export type IShoulds<I extends TestTypeParams_any> = Record<string, BaseShould<I>>;

// Export the BaseShould class as named export
// export { BaseShould };
